docs: Improve developer guide

reggie-k opened the pull request:
Closes #23668 

A summary so that the PR is easier to review - huge thanks to @nitishfy @todaywasawesome @ppapapetrou76 @dudinea  who reviewed it even before this summary :) 

- `index.md` became the starting point for the entire documentation flow, it replaces the `quick-start` and has links to all other documents in the developer guide (except for tilt and remote debugging - follow-up PRs for this will come), so this aims to be the central place that contributors will look at
- `quick-start` content went to separate files, `quick-start` itself was removed
- Most of the existing content in developer docs was left unchanged, mostly restructured to support more intuitive flow. Minor additions were made to existing content for clarifications
- Merged considerations of local or virtualized chain from multiple places to a single one. Added some clarifications such as that a local toolchain is preferred for debugging in an IDE
- Merged `argo-ui` contrib section from multiple places to a single one
- The order of the needed steps is `development-environment`, `development-toolchain`, `development cycle`, `running Argo CD locally`, `debugging Argo CD locally`, `submitting your PR`
- Added prereqs for each such step 
- Created clear separation between the different local K8s clusters installations and their special config (Kind, K3d, Minikube)
- Added a dedicated section for Mac troubleshooting
- Added a new section for debugging locally (including sample config for VSCode and Goland)
- Added the `notifications-engine` dependency section
- Documented the 3 ways to run Argo CD locally - `make start-local`, `make run`, `goreman start`
- Fixed broken links to Extensions docs
- Merged instructions on how to install and configure local K8s clusters from multiple places to a single one
- Added a dedicated section for docs contributors, since that requires only minimal setup and does not require full-blown dev env

<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

todaywasawesome commented on docs/developer-guide/development-cycle.md:
```diff
@@ -0,0 +1,118 @@
+2. [Toolchain Guide](toolchain-guide.md)
+
+## Preface 
+When you have developed and possibly manually tested the code you want to contribute, you should ensure that everything will build correctly. Commit your changes to the local copy of your Git branch and perform the following steps, for each step the commands for both local and virtualized toolchain are available. 
```
```suggestion
When you have developed and possibly manually tested the code you want to contribute, you should ensure that everything builds correctly. Commit your changes locally and perform the following steps, for each step the commands for both local and virtualized toolchain are listed. 
```

todaywasawesome commented on docs/developer-guide/development-cycle.md:
```diff
@@ -0,0 +1,118 @@
+## Preface 
+When you have developed and possibly manually tested the code you want to contribute, you should ensure that everything will build correctly. Commit your changes to the local copy of your Git branch and perform the following steps, for each step the commands for both local and virtualized toolchain are available. 
+
+### Docker priviliges for virtualized chain users
```
```suggestion
### Docker priviliges for virtualized toolchain users
```

todaywasawesome commented on docs/developer-guide/development-cycle.md:
```diff
@@ -0,0 +1,118 @@
+### Docker priviliges for virtualized chain users
+[These instructions](toolchain-guide.md#docker-privileges) are relevant for most of the steps below 
+
+### Using Podman for virtualized chain users
```
```suggestion
### Using Podman for virtualized toolchain users
```

todaywasawesome commented on docs/developer-guide/development-cycle.md:
```diff
@@ -0,0 +1,118 @@
+## Development cycle steps
+### Set kubectl context to argocd namespace
+
+Set kubectl config to argocd namespace is required for all the following steps to succeed.
```
```suggestion
Setting kubectl config to the argocd namespace is required for these steps to succeed.
```

todaywasawesome commented on docs/developer-guide/development-cycle.md:
```diff
@@ -0,0 +1,118 @@
+
+* `make dep-ui` or `make dep-ui-local`
+
+Argo CD recently migrated to Go modules. Usually, dependencies will be downloaded on build time, but the Makefile provides two targets to download and vendor all dependencies:
```
```suggestion
Argo CD recently migrated to Go modules. Usually, dependencies will be downloaded at build time, but the Makefile provides two targets to download and vendor all dependencies:
```

todaywasawesome commented on docs/developer-guide/development-cycle.md:
```diff
@@ -0,0 +1,118 @@
+
+* Run `make lint` or `make lint-local` and observe any errors reported by the Linter
+* Fix any of the errors reported and commit to your local branch
+* Finally, after the Linter reports no errors anymore, run `git status` or `git diff` to check for any changes made automatically by Lint
```
```suggestion
* Finally, after the Linter reports no errors, run `git status` or `git diff` to check for any changes made automatically by Lint
```

todaywasawesome commented on docs/developer-guide/development-cycle.md:
```diff
@@ -0,0 +1,118 @@
+
+### Run end-to-end tests
+
+The final step is running the End-to-End testsuite, which makes sure that your Kubernetes dependencies are working properly. This will involve starting all the Argo CD components on your computer. The end-to-end tests consists of two parts: a server component, and a client component.
```
```suggestion
The final step is running the End-to-End testsuite, which ensures that your Kubernetes dependencies are working properly. This will involve starting all the Argo CD components on your computer. The end-to-end tests consists of two parts: a server component, and a client component.
```

todaywasawesome commented on docs/developer-guide/index.md:
```diff
@@ -1,25 +1,59 @@
-* [Running Argo CD Locally](running-locally/)
+## Contributing to Argo CD backend, frontend and documentation
+
+This guide shall help you in setting up your build & test environment, so that you can start developing and testing bug fixes and feature enhancements without having to make too much effort in setting up a local toolchain.
```
```suggestion
This guide will help you set up your build & test environment, so that you can start developing and testing bug fixes and feature enhancements without having to make too much effort in setting up a local toolchain.
```

todaywasawesome reviewed (changes requested):
Made some small grammar edits and I think it's good to go.

nitishfy commented on docs/developer-guide/development-cycle.md:
```diff
@@ -0,0 +1,118 @@
+* Commit any possible changes to your local Git branch, an appropriate commit message would be `Changes from codegen`, for example.
+
+!!!note
+    There are a few non-obvious assets that are auto-generated. You should not change the autogenerated assets, as they will be overwritten by a subsequent run of `make codegen`. Instead, change their source files. Prominent examples of non-obvious auto-generated code are `swagger.json` or the installation manifest YAMLs.
```
```suggestion
   There are a few non-obvious assets that are auto-generated. You should not edit these generated assets directly, as they will be overwritten by a subsequent run of make codegen. Instead, modify the source files from which they are generated such as Go types, annotations, or .proto files and then re-run the appropriate code generation command. Prominent examples of non-obvious auto-generated code are `swagger.json` or the installation manifest YAMLs.
```

> Instead, change their source files.

I was talking to @crenshaw-dev the other day about it, and he mentioned that it actually depends on whether you should be changing the source files or the proto files. The above note is true wrt to `swagger.json` that should never be edited manually. However, If a `.proto` file generates the auto-generated asset, changing the proto file makes sense here. If a Go struct with annotations (like Kubernetes CRD definitions) is the source, changing the Go source file makes sense. Hence, it actually depends.

nitishfy replied:
Secondly, we should mention this detail about never changing the proto field number in case of field deprecation, as that may lead to API API-breaking change.

reggie-k replied:
I think we should deal with the codegen -related docs in a separate PR, this PR is more about restructuring stuff related to setting up the environment.

nitishfy commented on docs/developer-guide/debugging-locally.md:
```diff
@@ -0,0 +1,129 @@
+Below are the different options.
+
+### Run the other components locally
+#### Run with "make start-local"
```
```suggestion
#### Run with "make start-local"
```
Can we also mention about goreman here?

reggie-k replied:
@nitishfy I added in the preface a section that links to starting local services with local toolchain, and there there are more details about Goreman, is this enough or do you want me to link it also from this section?

nitishfy replied:
> Yeah, anything particular you had in mind?

I typically use goreamn while the other component is running on IDE. I think we can simply mention about `goreman run start <component-name>` command that makes sure the other components are run by goreman.

reggie-k replied:
So that is a third way of running local env? I never used it. I will try it first.

reggie-k replied:
For now, the PR has docs for all the 3 options for local runs (including Goreman, thanks for that!), let's continue the discussion on which one is recommended and when separately?

nitishfy reviewed (changes requested):
@reggie-k This is a solid PR! I've left some comments PTAL.

reggie-k commented:
> I forgot to mention the use of tilt here. Should we add this in the same PR (which is recommended) or carry its docs in another PR? There's already an [issue to add docs for tilt](https://github.com/argoproj/argo-cd/issues/23224).

I'd rather deal with Tilt separately, I think it would be easier once this baseline is established.

nitishfy commented on docs/developer-guide/debugging-locally.md:
```diff
@@ -0,0 +1,131 @@
+`make start-local` runs all the components by default, but it is also possible to run it with a whitelist of components, enabling the separation we need.
+
+So for the case of debugging the `api-server`, run:
+`make start-local ARGOCD_START="notification applicationset-controller repo-server redis dex controller ui"` 
```
I always wonder how you would determine which other components are required to be running while trying to run another component in debug mode. For example, to run the application controller in debug mode, you need to make sure `repo-server` and `redis` are running; hence, `make start-local ARGOCD_START="redis repo-server"` works. Having said that, we can provide [a link to the dependency chart](https://argo-cd.readthedocs.io/en/latest/developer-guide/architecture/components/#dependencies) so users can figure out of their own.

reggie-k replied:
I've removed the option of running partially in a cluster, it is rather complex and better for more advanced users.
We can add docs about this later on, this may require updating the architecture chart. Good catch!

dudinea commented on docs/developer-guide/debugging-locally.md:
```diff
@@ -0,0 +1,131 @@
+
+When running Argo CD locally for manual tests, the quickest way to do so is to run all the Argo CD components together, as described in [Running Locally](running-locally.md), 
+
+However, when you need to debug a single Argo CD component (for example, api-server, repo-server, etc), you will need to run this component separately in your IDE, using your IDE launch and debug configuration, while the other components will be running as described previously, using the local toolchain.
```
One does not always need to run separately in the IDE the component being debugged. One can attach with debuggers to already running processes, which were started in a regular way, like "make run"

reggie-k commented:
Executive summary so that the PR is easier to review:

- `index.yaml` became the starting point for the entire documentation flow, it replaces the `quick-start` and has links to all other documents in the developer guide (except for tilt and remote debugging - follow-up PRs for this will come), so this aims to be the central place that developers will look at
- `quick-start` content went to separate files, `quick-start` itself was removed
- Most of the existing content in developer docs was left unchanged, mostly restructured to support more intuitive flow. Minor additions were made to existing content for clarifications
- Merged considerations of local or virtualized chain from multiple places to a single one. Added some clarifications such as that a local toolchain is preferred for debugging in an IDE
- Merged `argo-ui` contrib section from multiple places to a single one
- The order of the needed steps is `development-environment`, `development-toolchain`, `development cycle`, `running Argo CD locally`, `debugging Argo CD locally`, `submitting your PR`
- Added prereqs for each such step 
- Created clear separation between the different local K8s clusters installations and their special config (Kind, K3d, Minikube)
- Added a dedicated section for Mac troubleshooting
- Added a new section for debugging locally (including sample config for VSCode and Golang)
- Documented the 3 ways to run Argo CD locally - `make start-local`, `make run`, `goreman start`
- Fixed broken links to Extensions docs
- Merged instructions on how to install and configure local K8s clusters rom multiple places to a single one

nitishfy commented on docs/developer-guide/debugging-locally.md:
```diff
@@ -0,0 +1,134 @@
+## Configure your IDE
+
+### Locate your component configuration in `Procfile`
+The `Profile` is used by Goreman when running Argo CD locally with the local toolchain. It has all the needed component run configuration, and you will need to copy parts of this configuration to your IDE.
```
```suggestion
The `Procfile` is used by Goreman when running Argo CD locally with the local toolchain. It has all the needed component run configuration, and you will need to copy parts of this configuration to your IDE.
```
nit

nitishfy commented on docs/developer-guide/debugging-locally.md:
```diff
@@ -0,0 +1,134 @@
+
+## Configure your IDE
+
+### Locate your component configuration in `Procfile`
```
```suggestion
### Locate your component configuration in `Procfile`
```
can we add a link to the Procfile here?

reggie-k commented:
One more direction I am thinking about is - we have 2 types of contributors:
Those who make quick fixes in the docs - they don't necessarily need a full blown Argo CD dev env for that.
Those who fix bugs or develop new features - they do need a full blown Argo CD dev env.

Currently, the PR is oriented for the second category. Maybe we need to have in addition a dedicated guide, which will be much shorter, for just making quick docs contributions?

nitishfy commented on docs/developer-guide/debugging-locally.md:
````diff
@@ -0,0 +1,134 @@
+#### Goland example
+Next, you will need to create a launch configuration, with the relevant parameters. Copy the parameters from `Procfile`, located in the `argo-cd` root folder of your development branch. The parameters are located after the `$COMMAND` section in the `sh -c` section of the component run command.
+Example for an `api-server` launch configuration, based on our above example for `api-server` configuration in `Procfile`: 
+``` xml 
````
Should we mention that? Otherwise, I think a better option would be to redirect users to the [GoLand debugging docs](https://www.jetbrains.com/help/go/debugging-code.html).

todaywasawesome commented on mkdocs.yml:
```diff
@@ -203,14 +203,18 @@ nav:
   - Code Contribution Guide: developer-guide/code-contributions.md
   - Toolchain Guide: developer-guide/toolchain-guide.md
-  - developer-guide/contributors-quickstart.md
+  - developer-guide/mac-users.md
```
I'm fine either way. It's nice to have the other pages a little shorter to put this in a corner.

reggie-k replied:
@nitishfy there are more issues with Mac, I don't have them listed right now, and also we will probably have a similar  file later on for Windows with WSL. I think it is important to have a dedicated section because this way it is easier for people to find their OS-specific issues.

nitishfy commented on docs/developer-guide/debugging-locally.md:
````diff
@@ -0,0 +1,143 @@
+Example for `api-server` configuration in `Procfile`:
+``` text
+api-server: [ "$BIN_MODE" = 'true' ] && COMMAND=./dist/argocd || COMMAND='go run ./cmd/main.go' && sh -c "GOCOVERDIR=${ARGOCD_COVERAGE_DIR:-/tmp/coverage/api-server} FORCE_LOG_COLORS=1 ARGOCD_FAKE_IN_CLUSTER=true ARGOCD_TLS_DATA_PATH=${ARGOCD_TLS_DATA_PATH:-/tmp/argocd-local/tls} ARGOCD_SSH_DATA_PATH=${ARGOCD_SSH_DATA_PATH:-/tmp/argocd-local/ssh} ARGOCD_BINARY_NAME=argocd-server $COMMAND --loglev ...
+```
````
```suggestion
```
!!! note
The Procfile for a component can get changed with time. Please go through the Procfile and make sure you use the latest configuration for debugging
!!!
```

nitishfy replied:
Let's add a note here for user to instruct them to use the latest content from the Procfile.

nitishfy commented on docs/developer-guide/debugging-locally.md:
````diff
@@ -0,0 +1,143 @@
+```
+
+#### Goland example
+Next, you will need to create a launch configuration, with the relevant parameters. Copy the parameters from `Procfile`, located in the `argo-cd` root folder of your development branch. The parameters are located after the `$COMMAND` section in the `sh -c` section of the component run command.
````
```suggestion
You'lll need to create a launch configuration, with the relevant parameters. Copy the parameters from `Procfile`, located in the `argo-cd` root folder of your development branch. The parameters are located after the `$COMMAND` section in the `sh -c` section of the component run command.
```

nitishfy commented on docs/developer-guide/debugging-locally.md:
```diff
@@ -0,0 +1,143 @@
+
+### Run the other components locally
+#### Run with "make start-local"
+`make start-local` runs all the components by default, but it is also possible to run it with a whitelist of components, enabling the separation we need.
```
LGTM. I just noticed that `make start-local` [under the hood use goreman](https://github.com/argoproj/argo-cd/blob/master/Makefile#L537) to start the components locally :)))

nitishfy commented on docs/developer-guide/debugging-locally.md:
```diff
@@ -0,0 +1,143 @@
+`make run exclude=api-server` 
+
+#### Run with "goreman start"
+`goreman start` runs all the components by default, but it is also possible to run it with a whitelist of components, enabling separation as needed.
```
```suggestion
`goreman start` runs all the components by default, but it is also possible to run it with a whitelist of components, enabling separation as needed. For example, `goreman run start controller repo-server....`
```

nitishfy commented on docs/developer-guide/debugging-locally.md:
```diff
@@ -0,0 +1,143 @@
+
+As an alternative to importing the above file to Goland, you can create a Run/Debug Configuration using the official [Goland docs](https://www.jetbrains.com/help/go/go-build.html) and just copy the `parameters`, `directory` and `PATH` sections from the example above (specifying `Run kind` as `Directory` in the Run/Debug Configurations wizard)
+
+## Run Argo CD without the debugged component
```
Have me mentioned it somewhere that we are required to scale down all the components already running in our cluster?

todaywasawesome reviewed (approved):
Good to go, any of the other comments can be resolved in an additional PR as they are not blocking but rather looking for potential additions.

reggie-k commented:
/cherry-pick release-3.1

reggie-k commented:
/cherry-pick release-3.1
